### PR TITLE
committing for review by Jarrod

### PR DESCRIPTION
**Pull request by smith-sj**

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -0,0 +1,2 @@
+source 'https://rubygems.org'
+gem 'cspec'
```
Comment (jarroddalefolino) [inline Gemfile]: Add rubocop and run the autofix `rubocop -a` https://github.com/rubocop/rubocop

File: adjudicator.rb [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+            (current_game[0][0][0] + current_game[0][1][0] + current_game[0][2][0]),
+            (current_game[0][0][1] + current_game[0][1][1] + current_game[0][2][1]),
+            (current_game[0][0][2] + current_game[0][1][2] + current_game[0][2][2]),
+            (current_game[0][0][0] + current_game[0][1][1] + current_game[0][2][2]),
+            (current_game[0][0][2] + current_game[0][1][1] + current_game[0][2][0]),
+            # board 2
+            (current_game[1][0][0] + current_game[1][0][1] + current_game[1][0][2]),
+            (current_game[1][1][0] + current_game[1][1][1] + current_game[1][1][2]),
+            (current_game[1][2][0] + current_game[1][2][1] + current_game[1][2][2]),
+            (current_game[1][0][0] + current_game[1][1][0] + current_game[1][2][0]),
```
Comment (jarroddalefolino) [inline adjudicator.rb]: we need to do this in a nested loop

File: adjudicator.rb [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+            end
+        end
+
+        diamond_stairs.each do |combo|
+            if combo == combo_key[0]
+                @tally[0][:x_tally] += diamond_stairs_points
+            elsif combo == combo_key[1]
+                @tally[1][:o_tally] += diamond_stairs_points
+            end
+        end
```
Comment (jarroddalefolino) [inline adjudicator.rb]: same here, loop through an array `[pillars, stairs, diamond_stairs]` and reuse the logic

File: adjudicator.rb [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+        return @tally
+    end
+
+    def get_x_tally(game_squares)
+        return tally_up(game_squares)[0][:x_tally]
```
Comment (jarroddalefolino) [inline adjudicator.rb]: return is optional

File: adjudicator.rb [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+                @tally[0][:x_tally] += diamond_stairs_points
+            elsif combo == combo_key[1]
+                @tally[1][:o_tally] += diamond_stairs_points
+            end
+        end
+
+        return @tally
+    end
+
+    def get_x_tally(game_squares)
```
Comment (jarroddalefolino) [inline adjudicator.rb]: ruby never has methods that start with `get_`

File: ai.rb [hunk 6]
```diff
@@ -0,0 +1,80 @@
+require_relative "game_output"
+require_relative "adjudicator"
+
+class Ai
+
+    # returns a random name from list
+    def name
+        names = ['Bob', 'Jenny', 'HorseyChick93', 'SpicyKing22', 'Albert', 'ThePun!sh3r', 'Kyle', 'PJ', 'David', 'Samantha', 'Charlie', 'Keegan', 'TheGodfather', 'Janice']
```
Comment (jarroddalefolino) [inline ai.rb]: Make this a constant

File: ai.rb [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+        return names[rand(names.length)]
+    end
+    
+    # returns a random move, i.e. "2c"
+    def get_move
+        letters = ["a","b","c"]
```
Comment (jarroddalefolino) [inline ai.rb]: move to a constant

File: game_board.rb [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+            [" "," "," "]],
+            # board 2
+            [[" "," "," "],
+            [" "," "," "],
+            [" "," "," "]],
+            # baord 3
+            [[" "," "," "],
+            [" "," "," "],
+            [" "," "," "]],
+        ]
```
Comment (jarroddalefolino) [inline game_board.rb]: generate with a nested iteration

File: game_input.rb [hunk 9]
```diff
@@ -0,0 +1,14 @@
+class GameInput
+
+    # rearranges a2 => 2a etc.
+    def reorder_move(move)
+        return move.chars.sort.join
+    end
```
Comment (jarroddalefolino) [inline game_input.rb]: This is called `GameInput` yet the method here does no input. It is formatting output

File: game_modes.rb [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+                    elsif x_final_tally < o_final_tally
+                        puts "\nGame over! #{player_o_name} wins!"
+                    else
+                        puts "\nGame over! It's a draw."
+                    end
+                    break
+                end
+            else
+                Player.new.play_turn(player_x_name, player_o_name, player_name, token, game_squares)
+                if turn_count >= 27
```
Comment (jarroddalefolino) [inline game_modes.rb]: what is `27`. A magic number? Move it to a constant and give it a meaningful name

File: game_validator.rb [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+    VALID_MOVES = [
+        "1a", "1b", "1c", 
+        "2a", "2b", "2c", 
+        "3a", "3b", "3c",
+        "4a", "4b", "4c", 
+        "5a", "5b", "5c", 
+        "6a", "6b", "6c",
+        "7a", "7b", "7c", 
+        "8a", "8b", "8c", 
+        "9a", "9b", "9c"]
```
Comment (jarroddalefolino) [inline game_validator.rb]: single quotes unless needed

File: main.rb [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+
+        # Player vs AI
+
+        elsif game_mode == "2"
+            game_board = GameBoard.new
+            game_squares = game_board.squares
+            GameMode.new.p_v_ai(game_squares)
+            puts "Press ENTER to continue"
+            gets
+            break
```
Comment (jarroddalefolino) [inline main.rb]: we need to get rid of the repeated logic

File: specs.csv [hunk 13]
```diff
@@ -0,0 +1,7 @@
+class,id,name,type,method,initialization_arg_1,method_arg_1,method_arg_2,method_arg_3,expected
+GameInput,GameInput-1.1,re-order move from a1->1a,instance,reorder_move,,a1,,,"1a"
+TurnMaster,Turnmaster-1.1,Checks who's turn it is,instance,whos_turn,,1,"Steve","Bob","Steve"
+Adjudicator,Adjudicator-1.2,Getting x score of 0,instance,get_x_tally,,"[[[' ',' ',' '],[' ',' ',' '],[' ',' ',' ']],[[' ',' ',' '],[' ',' ',' '],[' ',' ',' ']],[[' ',' ',' '],[' ',' ',' '],[' ',' ',' ']]]",,,"0"
+Adjudicator,Adjudicator-1.3,Getting o score of 0,instance,get_o_tally,,"[[[' ',' ',' '],[' ',' ',' '],[' ',' ',' ']],[[' ',' ',' '],[' ',' ',' '],[' ',' ',' ']],[[' ',' ',' '],[' ',' ',' '],[' ',' ',' ']]]",,,"0"
+Adjudicator,Adjudicator-1.4,Getting x score of 20,instance,get_x_tally,,"[[['x','o','x'],['x','x','o'],['o','o','x']],[['x','x','x'],['o','o','o'],['o','o','o']],[['x','o','x'],['x','o','x'],['x','o','x']]]",,,"20"
+Adjudicator,Adjudicator-1.5,Getting o score of 15,instance,get_o_tally,,"[[['x','o','x'],['x','x','o'],['o','o','x']],[['x','x','x'],['o','o','o'],['o','o','o']],[['x','o','x'],['x','o','x'],['x','o','x']]]",,,"15"
```
Comment (jarroddalefolino) [inline specs.csv]: good

File: player.rb [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+            # if the move is not valid, restart loop
+            elsif !GameValidator.new.is_valid(move)
+                GameOutput.new.print_game(player_x, player_o, game_squares, x_tally, o_tally)
+                puts "\nInvalid Move! Try again.\n"
+            # if the move is not unique, check on next board
+            elsif !GameValidator.new.is_unique(move, game_squares)
+                # adds 3 to row (2a -> 5a) same spot on next board
+                move[0..0]= ((move[0]).to_i + 3).to_s
+                # if the new move is unique add it to second board
+                if GameValidator.new.is_unique(move, game_squares)
```
Comment (jarroddalefolino) [inline player.rb]: in general we want to get rid of nested if statements

File: main.rb [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+            
+        # AI vs AI
+
+        elsif game_mode == "3"
+            game_board = GameBoard.new
+            game_squares = game_board.squares
+            GameMode.new.ai_v_ai(game_squares)
+            puts "Press ENTER to continue"
```
Comment (jarroddalefolino) [inline main.rb]: `"Press ENTER to continue"` is repeated. Move to a constant

File: game_modes.rb [hunk 16]
```diff
@@ -0,0 +1,146 @@
...
+
+        turn_count = 0
+
+        while true
+            turn_count += 1
+            player_name = TurnMaster.new.whos_turn(turn_count, player_x_name, player_o_name)
+            token = player_name == "#{player_x_name}" ? "x" : "o"
+            if player_name == player_o_name
+                Ai.new.play_ai_turn(player_x_name, player_o_name, player_name, token, game_squares)
+                if turn_count >= 27
```
Comment (jarroddalefolino) [inline game_modes.rb]: Magic number

**Review by jarroddalefolino — CHANGES_REQUESTED** [review]

File: Gemfile (hunk 1 above)
Comment (smith-sj) [reply Gemfile]: done

File: adjudicator.rb (hunk 2 above)
Comment (smith-sj) [reply adjudicator.rb]: done

File: adjudicator.rb (hunk 3 above)
Comment (smith-sj) [reply adjudicator.rb]: done

File: adjudicator.rb (hunk 4 above)
Comment (smith-sj) [reply adjudicator.rb]: done

File: adjudicator.rb (hunk 5 above)
Comment (smith-sj) [reply adjudicator.rb]: fixed

File: main.rb (hunk 12 above)
Comment (smith-sj) [reply main.rb]: done

File: specs.csv (hunk 13 above)
Comment (smith-sj) [reply specs.csv]: thanks - I'll add more

File: main.rb (hunk 15 above)
Comment (smith-sj) [reply main.rb]: done

File: game_modes.rb (hunk 16 above)
Comment (smith-sj) [reply game_modes.rb]: fixed

File: player.rb (hunk 14 above)
Comment (smith-sj) [reply player.rb]: done

File: ai.rb (hunk 6 above)
Comment (smith-sj) [reply ai.rb]: done

File: ai.rb (hunk 7 above)
Comment (smith-sj) [reply ai.rb]: done

File: game_input.rb (hunk 9 above)
Comment (smith-sj) [reply game_input.rb]: moved to game_validator

File: game_modes.rb (hunk 10 above)
Comment (smith-sj) [reply game_modes.rb]: assigned to MAX_TURNS

File: game_validator.rb (hunk 11 above)
Comment (smith-sj) [reply game_validator.rb]: done

File: game_board.rb (hunk 8 above)
Comment (smith-sj) [reply game_board.rb]: done

**Pull request merged**